components/playlists/browser/playlists_controller.cc

```diff
@@ -0,0 +1,539 @@
+
+}
+
+bool PlaylistsController::Init(const base::FilePath& base_dir) {
+  base_dir_ = base_dir;
+  db_controller_.reset(
+      new PlaylistsDBController(base_dir.Append(kDatabaseDirName)));
+  media_file_controller_.reset(
+      new PlaylistsMediaFileController(this, url_loader_factory_));
+
+  return base::PostTaskAndReplyWithResult(
+      io_task_runner(),
+      FROM_HERE,
+      base::BindOnce(&PlaylistsDBController::Init,
+                     base::Unretained(db_controller_.get())),
+      base::BindOnce(&PlaylistsController::OnDBInitialized,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void PlaylistsController::OnDBInitialized(bool initialized) {
+  initialized_ = initialized;
+  LOG(ERROR) << __func__ << ": " << initialized_;
+
+  if (!initialized_) {
+    return;
+  }
+
+  for (PlaylistsControllerObserver& obs : observers_)
+    obs.OnPlaylistsInitialized();
+}
+
+void PlaylistsController::NotifyPlaylistChanged(
+    const PlaylistsChangeParams& params) {
+  for (PlaylistsControllerObserver& obs : observers_)
+    obs.OnPlaylistsChanged(params);
+}
+
+void PlaylistsController::DownloadThumbnail(base::Value&& playlist_value,
+                                            bool directory_ready) {
+  if (!directory_ready)
+    return;
+
+  base::Value value = std::move(playlist_value);
+  const std::string* thumbnail_url =
+      value.FindStringPath("createParams.playlistThumbnailUrl");
+
+  if (!thumbnail_url || (*thumbnail_url).empty()) {
+    // TODO(simonhong): If thumbnail url is not available skip downloading it
+    // and goes step 2 directly.
+    NOTREACHED() << " thumbnail url is not available";
+  }
+
+  auto request = std::make_unique<network::ResourceRequest>();
+  request->url = GURL(*thumbnail_url);
+  request->allow_credentials = false;
+  network::SimpleURLLoader* loader = network::SimpleURLLoader::Create(
+      std::move(request),
+      GetNetworkTrafficAnnotationTagForURLLoad()).release();
+  loader->SetAllowHttpErrorResults(true);
+  url_loaders_.insert(loader);
+  loader->SetRetryOptions(
+      kRetriesCountOnNetworkChange,
+      network::SimpleURLLoader::RetryMode::RETRY_ON_NETWORK_CHANGE);
+
+  const std::string playlist_id = *value.FindStringKey("id");
+  base::FilePath thumbnail_path =
+      base_dir_.Append(GetPlaylistIDDirName(playlist_id)).
+          Append(kThumbnailFileName);
+  loader->DownloadToFile(
+      url_loader_factory_.get(),
+      base::BindOnce(&PlaylistsController::OnThumbnailDownloaded,
+                     base::Unretained(this),
+                     std::move(value),
+                     loader),
+      thumbnail_path);
+}
+
+void PlaylistsController::OnThumbnailDownloaded(
+    base::Value&& playlist_value,
+    network::SimpleURLLoader* loader,
+    base::FilePath path) {
+  DCHECK(url_loaders_.find(loader) != url_loaders_.end());
+  url_loaders_.erase(loader);
+  std::unique_ptr<network::SimpleURLLoader> scoped_loader(loader);
+
+  if (path.empty())
```

Comment:
at this point the thumbnail download has failed, and we won't ever fill in the `thumbnailPath` value in the playlist record, and we don't notify anyone that the download failed and that the `thumbnailPath` value is never going to happen. is that a problem?

Reply:
When thumbnail is not available, will move to next step(media file generation).
and extension will get thumbnail fail noti.
Then, extension can recover/retry fetching again by using new `recoverPlaylist` api.